Src/Newtonsoft.Json/Utilities/TypeExtensions.cs

```diff
@@ -273,6 +273,226 @@ public static bool IsDefined(this Type type, Type attributeType, bool inherit)
+                return false;
+            }
+
+            // check if the last parameter is ParamArray
+            Type paramArrayType = null;
+            
+            if (enableParamArray &&
+                parameters[parameters.Length - 1].ParameterType.IsArray &&
+                parameters[parameters.Length - 1].IsDefined(typeof(ParamArrayAttribute)))
+            {
+                paramArrayType = parameters[parameters.Length - 1].ParameterType.GetElementType();
+            }
+
+            if (paramArrayType == null && parameters.Length != types.Count)
+            {
+                // when there's no ParamArray, number of parameters should match
+                return false;
+            }
+            
+            for (int i = 0; i < types.Count; i++)
+            {
+                var paramType = (paramArrayType != null && i >= parameters.Length - 1) ? paramArrayType : parameters[i].ParameterType;
+
+                // exact match with provided type
+                if (paramType == types[i]) continue;
+
+                // parameter of type object matches anything
+                if (paramType == typeof(object)) continue;
+
+                if (paramType.IsPrimitive())
+                {
+                    // primitive parameter can only be assigned from compatible primitive type
+                    if (!types[i].IsPrimitive() || !CanConvertPrimitive(types[i], paramType))
+                        return false;
+                }
+                else
+                {
+                    if (!paramType.IsAssignableFrom(types[i]))
+                        return false;
+                }
+            }
+
+            return true;
+        }
+
+        private class ParametersMatchComparer : IComparer<ParameterInfo[]>
+        {
+            private readonly IList<Type> _types;
+            private readonly bool _enableParamArray;
+
+            public ParametersMatchComparer(IList<Type> types, bool enableParamArray)
+            {
+                if (types == null)
+                    throw new ArgumentNullException(nameof(types));
+
+                _types = types;
+                _enableParamArray = enableParamArray;
+            }
+
+            public int Compare(ParameterInfo[] parameters1, ParameterInfo[] parameters2)
+            {
+                if (parameters1 == null)
+                    throw new ArgumentNullException(nameof(parameters1));
+                if (parameters2 == null)
+                    throw new ArgumentNullException(nameof(parameters2));
+
+                // parameterless method wins
+                if (parameters1.Length == 0) return -1;
+                if (parameters2.Length == 0) return 1;
+
+                Type paramArrayType1 = null, paramArrayType2 = null;
+
+                if (_enableParamArray)
+                {
+                    if (parameters1[parameters1.Length - 1].ParameterType.IsArray &&
```

Comment:
Assign `parameters1[parameters1.Length - 1]` to a local variable rather than accessing the array multiple times